Introduce `-Zvirtual-function-elimination` codegen flag

flip1995 opened the pull request:
Fixes #68262

This PR adds a codegen flag `-Zvirtual-function-elimination` to enable the VFE optimization in LLVM. To make this work, additonal  information has to be added to vtables ([`!vcall_visibility` metadata](https://llvm.org/docs/TypeMetadata.html#vcall-visibility-metadata) and a `typeid` of the trait). Furthermore, instead of just `load`ing functions, the [`llvm.type.checked.load` intrinsic](https://llvm.org/docs/LangRef.html#llvm-type-checked-load-intrinsic) has to be used to map functions to vtables.

For technical details of the changes, see the commit messages.

I also tested this flag on https://github.com/tock/tock on different boards to verify that this fixes the issue https://github.com/tock/tock/issues/2594. This flag is able to improve the size of the resulting binary by about 8k-9k bytes by removing the unused debug print functions.

[Rendered documentation update](https://github.com/flip1995/rust/blob/pk-vfe/src/doc/rustc/src/codegen-options/index.md#virtual-function-elimination)

rust-highfive commented:
Some changes occured to rustc_codegen_gcc

cc @antoyo

rust-highfive commented:
r? @nagisa

(rust-highfive has picked a reviewer for you, use r? to override)

flip1995 commented:
I think this flag fits better as a `-C` flag because it influences codegen and is not a debugging/nightly only flag. I can change it to a `-Z` flag, if it should start out as a nightly only flag though.

Mark-Simulacrum commented:
Is there a reason why we shouldn't (or LLVM shouldn't) enable this optimization by default with full LTO? Is there an expectation that it might miscompile something or otherwise cause us unexpected trouble?

tmiasko commented on compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs:
```diff
@@ -1433,6 +1435,84 @@ fn build_vtable_type_di_node<'ll, 'tcx>(
+    let trait_ref_self = trait_ref.with_self_ty(cx.tcx, ty);
+    let trait_ref_self = cx.tcx.erase_regions(trait_ref_self);
+    let trait_def_id = trait_ref_self.def_id();
+    let trait_vis = cx.tcx.visibility(trait_def_id);
```
I think that the closest existing concept similar  to `vcall_visibility` would be reachability, rather than visibility.

<details><summary>Example with an incorrectly optimized out vtables</summary>

```rust
trait Foo { fn foo(&self) { println!("foo") } }
trait Bar { fn bar(&self) { println!("bar") } }
trait Baz { fn baz(&self) { println!("baz") } }

impl Foo for usize {}
impl Bar for usize {}
impl Baz for usize {}

pub struct FooBox(Box<dyn Foo>);
pub struct BarBox(Box<dyn Bar>);
pub struct BazBox(Box<dyn Baz>);

pub fn make_foo() -> FooBox { FooBox(Box::new(0)) }
pub fn make_bar() -> BarBox { BarBox(Box::new(0)) }
pub fn make_baz() -> BazBox { BazBox(Box::new(0)) }

#[inline]
pub fn f(a: FooBox) { a.0.foo() }
pub fn g<T>(b: BarBox) { b.0.bar() }
#[inline]
fn h(c: BazBox) { c.0.baz() }
pub const H: fn(BazBox) = h;
```

</details>

flip1995 replied:
Thanks for pointing this out. I don't really have a solution for this right now. I tried looking at the reachability of the `trait_ref` and the members of the traits with the [`privacy_access_levels`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.privacy_access_levels) query. But for all of the traits and the functions in the traits (queried with `tcx.vtable_entries`) this just returns `None`, meaning it is not reachable IIUC.

I guess I would somehow have to check the reachability for the `dyn Trait` types. But I couldn't figure out how to do that.

Do you have ideas on how to best address this issue?

nikic commented:
> I think this flag fits better as a `-C` flag because it influences codegen and is not a debugging/nightly only flag. I can change it to a `-Z` flag, if it should start out as a nightly only flag though.

Yes, new flags are *always* added as unstable `-Z` flags first, and may later be stabilized as a `-C` flag. The "debugging options" terminology is an anachronism, the actual distinction between `-C` and `-Z` is stability.

bjorn3 commented on compiler/rustc_codegen_llvm/src/back/lto.rs:
```diff
@@ -591,6 +591,18 @@ pub(crate) fn run_pass_manager(
+                module.module_llvm.llmod(),
+                llvm::LLVMModFlagBehavior::Error,
+                "LTOPostLink\0".as_ptr().cast(),
+                1,
```
Should this be 11?

nagisa replied:
I believe the `HasModuleFlag` should be checking for `1`, but also adding this flag here seems super sketchy to me. This is a pretty internal thing to LTO AFAICT.

In that context, it would be great if this code had a comment with context as to why this is being done.

tschuett replied:
In the mono repo, I found the value 1.

flip1995 replied:
The code here is based on the file tschuett linked above. In the LLVM repo that code is responsible to set up LTO. Adding this flag was missing from the Rust code that should emulate the code in `LTOCodeGenerator.cpp`.

The `11` argument in the `HasModuleFlag` is the size of the string passed to this function, not the value of the flag. The `1` argument in the `AddModuleFlag` function is the value of the module flag. LLVM doesn't need a string size here, because it expects a `\0` terminated string in this function.

bjorn3 commented on compiler/rustc_symbol_mangling/src/v0.rs:
```diff
@@ -93,6 +93,26 @@ pub(super) fn mangle_typeid_for_fnabi<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    trait_ref: ty::PolyExistentialTraitRef<'tcx>,
+) -> String {
+    // FIXME(flip1995): Seem comment in `mangle_typeid_for_fnabi`.
```
*See

bjorn3 commented on compiler/rustc_symbol_mangling/src/v0.rs:
```diff
@@ -93,6 +93,26 @@ pub(super) fn mangle_typeid_for_fnabi<'tcx>(
+    trait_ref: ty::PolyExistentialTraitRef<'tcx>,
+) -> String {
+    // FIXME(flip1995): Seem comment in `mangle_typeid_for_fnabi`.
+    let prefix = "_ZTS";
```
I don't think this needs a specific prefix. Rust vtables aren't compatible with C++ anyway.

bjorn3 commented on compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs:
```diff
@@ -1433,6 +1435,84 @@ fn build_vtable_type_di_node<'ll, 'tcx>(
+
+    let lto = cx.sess().lto();
+
+    // Since LLVM requires full LTO for the virtul function elimination optimization to apply, only
```
*virtual

nagisa commented on src/test/codegen/virtual-function-elimination.rs:
```diff
@@ -0,0 +1,99 @@
+    taking_v(&s);
+}
+
+// CHECK: ![[TYPE0]] = !{i64 0, !"[[MANGLED_TYPE0]]"}
```
Either this test or some other test should really be testing the exact values specified in the metadata here so that it can be seen if there are any regressions in under/overspecification caused by changes to the relevant code in the future.

nagisa commented on compiler/rustc_session/src/session.rs:
```diff
@@ -1419,14 +1419,17 @@ fn validate_commandline_args_with_session_available(sess: &Session) {
-            || sess.opts.cg.lto == config::LtoCli::Thin
-        {
+    // LLVM CFI and VFE both require LTO.
+    if sess.opts.cg.lto == config::LtoCli::Unspecified
```
This, I feel, should be using `sess.lto() == config::Lto::Fat` so that e.g. a specific target implying fat LTO is also taken into account here.

tschuett commented on compiler/rustc_codegen_llvm/src/back/lto.rs:
```diff
@@ -591,6 +591,18 @@ pub(crate) fn run_pass_manager(
     //      tools/lto/LTOCodeGenerator.cpp
```
I could not find tools/lto/LTOGenerator.cpp in the mono repo.

flip1995 commented:
Thanks for all the feedback. I think I addressed all the comments (I will clean up / squash the commits before this PR gets merged), except for the reachability issue, which needs a little bit more work/thinking on how to address this.

> Is there a reason why we shouldn't (or LLVM shouldn't) enable this optimization by default with full LTO? Is there an expectation that it might miscompile something or otherwise cause us unexpected trouble?

Clang/LLVM doesn't set this flag by default, so I figured rustc shouldn't set this flag by default either. I don't know if there are any known issues with this optimization in LLVM

nagisa commented:
This seems reasonable to me for a first shot at the functionality. Since it is unstable, I'm not super concerned about this being partially broken in some ways as e.g. pointed out by @tmiasko above, although we should definitely make sure the information about this limitation/bug is preserved in e.g. the tracking issue.

Another thing I'm not seeing is documentation for this feature the [unstable book](https://doc.rust-lang.org/nightly/unstable-book/compiler-flags.html). Please write something about this feature here, and especially please note the known limitations and/or issues as well.

flip1995 commented:
I added the documentation, cleaned up the commits and rebased on top of latest master.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6255760441?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/34aac0bc0570aebe64e5d2ee6a7a71c10d163d38/checks/6255760441/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_passes v0.0.0 (/checkout/compiler/rustc_passes)
    Checking rustc_save_analysis v0.0.0 (/checkout/compiler/rustc_save_analysis)
    Checking rustc_codegen_ssa v0.0.0 (/checkout/compiler/rustc_codegen_ssa)
    Checking rustc_resolve v0.0.0 (/checkout/compiler/rustc_resolve)
error[E0609]: no field `virtual_function_elimination` on type `CodegenOptions`
  --> compiler/rustc_codegen_ssa/src/meth.rs:28:35
   |
28 |         if bx.cx().sess().opts.cg.virtual_function_elimination && bx.cx().sess().lto() == Lto::Fat {
   |
   |
   = note: available fields are: `ar`, `code_model`, `codegen_units`, `control_flow_guard`, `debug_assertions` ... and 41 others
    Checking rustc_trait_selection v0.0.0 (/checkout/compiler/rustc_trait_selection)
For more information about this error, try `rustc --explain E0609`.
error: could not compile `rustc_codegen_ssa` due to previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6258976564?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/706c537c06efc7ca0943cfb4c814c78398622cec/checks/6258976564/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Rustbook (x86_64-unknown-linux-gnu) - embedded-book
Rustbook (x86_64-unknown-linux-gnu) - edition-guide
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.16s
unstable-book/print.html:1440: broken link fragment `#lto` pointing to `unstable-book/compiler-flags/virtual-function-elimination.html`
unstable-book/compiler-flags/virtual-function-elimination.html:146: broken link fragment `#lto` pointing to `unstable-book/compiler-flags/virtual-function-elimination.html`
number of HTML files scanned: 32297
number of HTML redirects found: 10094
number of links checked: 2249913
number of links ignored due to external: 112700
```

</details>

insilications commented:
Very nice work.

nagisa commented:
@bors r+ Thanks!

bors commented:
:pushpin: Commit 798a2d927f3796c0b21fc6ee2707ce21833bfbd0 has been approved by `nagisa`

<!-- @bors r=nagisa 798a2d927f3796c0b21fc6ee2707ce21833bfbd0 -->
<!-- homu: {"type":"Approved","sha":"798a2d927f3796c0b21fc6ee2707ce21833bfbd0","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit 798a2d927f3796c0b21fc6ee2707ce21833bfbd0 with merge 2c9812125037f96b4ba035183f237dfacc0f8d4a...
<!-- homu: {"type":"BuildStarted","head_sha":"798a2d927f3796c0b21fc6ee2707ce21833bfbd0","merge_sha":"2c9812125037f96b4ba035183f237dfacc0f8d4a"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6380115542?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/6380115542?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`i686-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/6379918051?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/2c9812125037f96b4ba035183f237dfacc0f8d4a/checks/6379918051/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Some tests failed in compiletest suite=codegen mode=codegen host=i686-unknown-linux-gnu target=i686-unknown-linux-gnu

failures:

---- [codegen] src/test/codegen/virtual-function-elimination.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/i686-unknown-linux-gnu/llvm/build/bin/FileCheck" "--input-file" "/checkout/obj/build/i686-unknown-linux-gnu/test/codegen/virtual-function-elimination/virtual-function-elimination.ll" "/checkout/src/test/codegen/virtual-function-elimination.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC"
stdout: none
--- stderr -------------------------------
/checkout/src/test/codegen/virtual-function-elimination.rs:66:12: error: CHECK: expected string not found in input
 // CHECK: @llvm.type.checked.load({{.*}}, i32 24, metadata !"[[MANGLED_TYPE0:[0-9a-zA-Z_]+]]")
           ^
/checkout/obj/build/i686-unknown-linux-gnu/test/codegen/virtual-function-elimination/virtual-function-elimination.ll:70:88: note: scanning from here
; <virtual_function_elimination::S as virtual_function_elimination::V>::public_function
                                                                                       ^
/checkout/obj/build/i686-unknown-linux-gnu/test/codegen/virtual-function-elimination/virtual-function-elimination.ll:82:29: note: possible intended match here
 %1 = tail call { i8*, i1 } @llvm.type.checked.load(i8* nonnull %0, i32 12, metadata !"NtCsfRpWlKdQPZn_28virtual_function_elimination1V")


Input file: /checkout/obj/build/i686-unknown-linux-gnu/test/codegen/virtual-function-elimination/virtual-function-elimination.ll
Check file: /checkout/src/test/codegen/virtual-function-elimination.rs

-dump-input=help explains the following input dump.
Input was:
<<<<<<
            .
            .
            .
            .
           65: define internal i32 @_RNvYNtCsfRpWlKdQPZn_28virtual_function_elimination1SNtB4_1U13subtrait_usedB4_(%S* noalias nocapture noundef nonnull readonly align 1 %self) unnamed_addr #1 { 
           67:  ret i32 4 
           68: } 
           69:  
           69:  
           70: ; <virtual_function_elimination::S as virtual_function_elimination::V>::public_function 
check:66'0                                                                                            X error: no match found
           71: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn 
check:66'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           72: define i32 @_RNvXs0_CsfRpWlKdQPZn_28virtual_function_eliminationNtB5_1SNtB5_1V15public_function(%S* noalias nocapture nonnull readonly align 1 %self) unnamed_addr #1 { 
check:66'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           73: start: 
check:66'0     ~~~~~~~
           74:  ret i32 6 
check:66'0     ~~~~~~~~~~~
           75: } 
check:66'0     ~~
           76:  
check:66'0     ~
           77: ; virtual_function_elimination::taking_v 
check:66'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           78: ; Function Attrs: nonlazybind uwtable 
check:66'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           79: define i32 @_RNvCsfRpWlKdQPZn_28virtual_function_elimination8taking_v({}* noundef nonnull align 1 %v.0, [3 x i32]* noalias noundef readonly align 4 dereferenceable(12) %v.1) unnamed_addr #2 { 
check:66'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           80: start: 
check:66'0     ~~~~~~~
           81:  %0 = bitcast [3 x i32]* %v.1 to i8* 
check:66'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           82:  %1 = tail call { i8*, i1 } @llvm.type.checked.load(i8* nonnull %0, i32 12, metadata !"NtCsfRpWlKdQPZn_28virtual_function_elimination1V") 
check:66'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:66'1                                 ?                                                                                                              possible intended match
           83:  %2 = extractvalue { i8*, i1 } %1, 0 
check:66'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           84:  %3 = bitcast i8* %2 to i32 ({}*)* 
check:66'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           85:  %4 = tail call i32 %3({}* noundef nonnull align 1 %v.0) 
check:66'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           86:  ret i32 %4 
check:66'0     ~~~~~~~~~~~~
           87: } 
check:66'0     ~~
            .
            .
>>>>>>
------------------------------------------
```

</details>

flip1995 commented on src/test/codegen/virtual-function-elimination.rs:
```diff
@@ -0,0 +1,99 @@
+}
+
+fn taking_t(t: &dyn T) -> i32 {
+    // CHECK: @llvm.type.checked.load({{.*}}, i32 24, metadata !"[[MANGLED_TYPE0:[0-9a-zA-Z_]+]]")
```
Those `i32 24` offsets in the `llvm.type.checked.load` are about half the value on 32-bit platforms, because pointer size is only 4-byte instead of 8-byte. How should I fix this? ignoring those tests on 32 bit architectures? Adding different `CHECK` comments for 32-bit platforms?

flip1995 commented:
I added `ignore-32bit` to the test and added a test with `ignore-64bit` to also test this on 32bit platforms.

nagisa commented:
@bors r+

bors commented:
:pushpin: Commit e2a5d02c75b02d817d026d577a50d966fde1459d has been approved by `nagisa`

<!-- @bors r=nagisa e2a5d02c75b02d817d026d577a50d966fde1459d -->
<!-- homu: {"type":"Approved","sha":"e2a5d02c75b02d817d026d577a50d966fde1459d","approver":"nagisa"} -->

klensy commented:
Title and description should talk about  `-Zvirtual-function-elimination`, not `-C` yes?

klensy commented on compiler/rustc_codegen_llvm/src/back/lto.rs:
```diff
@@ -588,9 +588,21 @@ pub(crate) fn run_pass_manager(
     unsafe {
+        if !llvm::LLVMRustHasModuleFlag(
+            module.module_llvm.llmod(),
+            "LTOPostLink".as_ptr().cast(),
```
Ahh, 11 is length of string. Nice API.

flip1995 replied:
> Nice API.

Yep the LLVM API for the module flags is pretty inconsistent...

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout pk-vfe` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self pk-vfe --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Auto-merging compiler/rustc_symbol_mangling/src/v0.rs
Auto-merging compiler/rustc_symbol_mangling/src/lib.rs
Auto-merging compiler/rustc_session/src/session.rs
Auto-merging compiler/rustc_codegen_llvm/src/intrinsic.rs
Auto-merging compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs
CONFLICT (content): Merge conflict in compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs
Auto-merging compiler/rustc_codegen_llvm/src/back/lto.rs
Auto-merging Cargo.lock
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

flip1995 commented:
> Title and description should talk about `-Zvirtual-function-elimination`, not `-C` yes?

Good catch! I updated the PR name and body :+1:

bors commented:
:umbrella: The latest upstream changes (presumably #97287) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nagisa commented:
@bors delegate+

Please write `@bors r=nagisa` once you rebase again.

bors commented:
:v: @flip1995 can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"nagisa","delegate":"flip1995"} -->

bors commented:
:pushpin: Commit fbfafd798a267ba90ef06d86cee666bd37b672a5 has been approved by `nagisa``

<!-- @bors r=nagisa` fbfafd798a267ba90ef06d86cee666bd37b672a5 -->
<!-- homu: {"type":"Approved","sha":"fbfafd798a267ba90ef06d86cee666bd37b672a5","approver":"nagisa`"} -->

nagisa commented:
@bors r- and here I hoped it would ignore the code quotes.

bors commented:
:hourglass: Testing commit 78e9608287dd3d4c974cefe98f95ef0b835b6347 with merge d84d7995bde5b58b4e18cdf1651f2c14adc15bb7...
<!-- homu: {"type":"BuildStarted","head_sha":"78e9608287dd3d4c974cefe98f95ef0b835b6347","merge_sha":"d84d7995bde5b58b4e18cdf1651f2c14adc15bb7"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6776355629?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/6776355629?check_suite_focus=true","builder_name":"checks-actions"} -->

bors commented:
:pushpin: Commit 195f2082002c9db456e0fde8c1d5db79929ae293 has been approved by `nagisa`

<!-- @bors r=nagisa 195f2082002c9db456e0fde8c1d5db79929ae293 -->
<!-- homu: {"type":"Approved","sha":"195f2082002c9db456e0fde8c1d5db79929ae293","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit 195f2082002c9db456e0fde8c1d5db79929ae293 with merge 2d1e0750792529248ed6f11061940c7203d668c9...
<!-- homu: {"type":"BuildStarted","head_sha":"195f2082002c9db456e0fde8c1d5db79929ae293","merge_sha":"2d1e0750792529248ed6f11061940c7203d668c9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6891003695?check_suite_focus=true)
Approved by: nagisa
Pushing 2d1e0750792529248ed6f11061940c7203d668c9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6891003695?check_suite_focus=true"},"merge_sha":"2d1e0750792529248ed6f11061940c7203d668c9"} -->

rust-timer commented:
Finished benchmarking commit (2d1e0750792529248ed6f11061940c7203d668c9): [comparison url](https://perf.rust-lang.org/compare.html?start=1f34da9ec8a85b6f86c5fa1c121ab6f88f2f4966&end=2d1e0750792529248ed6f11061940c7203d668c9&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=1f34da9ec8a85b6f86c5fa1c121ab6f88f2f4966&end=2d1e0750792529248ed6f11061940c7203d668c9&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1f34da9ec8a85b6f86c5fa1c121ab6f88f2f4966&end=2d1e0750792529248ed6f11061940c7203d668c9&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: mixed results
- Secondary benchmarks: mixed results

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 5.2%     | 5.2% | 1         |
| Regressions 😿 <br /> (secondary) | 4.5%     | 4.5% | 1         |
| Improvements 🎉 <br /> (primary) | -2.0%    | -2.0% | 1         |
| Improvements 🎉 <br /> (secondary) | -2.4%    | -2.4% | 1         |
| All 😿🎉 (primary) | 1.6%     | 5.2% | 2         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1f34da9ec8a85b6f86c5fa1c121ab6f88f2f4966&end=2d1e0750792529248ed6f11061940c7203d668c9&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
